File: code/com/leetcode/solutions/NO_0015_三数之和/Solution.js
/**
 * 思路：a + c + c = 0 且 a <= b <= c 
 *      => -a = b + c 且 a <= b <= c
 *      => a <= 0
 *      将数组排序后，设置下标 i, j = i+1, k = nums.length -1 (i < j < k)
 *      若 -nums[i] < num[j] + nums[k],这 k 左移
 *      若 -nums[i] > num[j] + nums[k],这 j 右移
 *      若 -nums[i] = num[j] + nums[k],返回[i,j,k],i左移并继续上诉操作
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    const ans = [];
    let sum, i = 0, prei;
    nums.sort((a, b) => a - b);
    for (let j, k, prej, prek; nums[i] < 0; i++) {
        if (nums[i] === nums[i - 1]) continue;
        j = i + 1;
        k = nums.length - 1;
        sum = null;
        while (j < k) {
            sum = nums[i] + nums[j] + nums[k]
            if (sum > 0) k--;
            else if (sum < 0) j++;
            else {
                if (nums[i] !== nums[prei] || nums[j] !== nums[prej] || nums[k] !== nums[prek]) {
                    ans.push([nums[i], nums[j], nums[k]]);
                }
                prei = i, prej = j, prek = k, k-- , j++;
            }
        }
    }
    if (nums[i] === 0 && nums[i + 1] === 0 && nums[i + 2] === 0) ans.push([nums[i], nums[i + 1], nums[i + 2]]);
    return ans;
};

File: code/com/leetcode/solutions/NO_0015_三数之和/Solution.java
package com.leetcode.solutions.NO_0015_三数之和;

import java.util.*;

/**
 * Created by gang.xu01@hand-china.com on 2019/1/19
 */
public class Solution {

    /**
     * 解法：将数组排序，定义三个指针，分别移动，使得前两个的和等于第三个的负数值，转化为两数之和
     *
     * @param nums 输入数组
     * @return
     */
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> list = new ArrayList<>();
        int len = nums.length;
        if (len < 3) return list;
        Arrays.sort(nums);
        int max = nums[len - 1];
        if (max < 0) return list;
        for (int i = 0; i < len - 2; ) {
            if (nums[i] > 0) break;
            if (nums[i] + 2 * max < 0) {
                while (nums[i] == nums[++i] && i < len - 2) ;
                continue;
            }
            int left = i + 1, right = len - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    list.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    while (nums[left] == nums[++left] && left < right) ;
                    while (nums[right] == nums[--right] && left < right) ;
                } else if (sum < 0) ++left;
                else --right;
            }
            while (nums[i] == nums[++i] && i < len - 2) ;
        }
        return list;
    }

    public static void main(String[] args) {
        System.out.println(new Solution().threeSum(new int[]{-5, 3, 2}));
    }

}
